### PR TITLE
[`pulse_generator`] Documentation (3/3) 

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
This is part 1 out of 3 into which we split #4073 for ease of reviewing.
See #4073 for all details regarding the overall PR.

**Description of the Change:**
Here we implement the `pulse_generator` gradient method, using the helper methods introduced in the first PR.
The parts missing are the changelog entry and the docstring of `pulse_generator`.

**Related GitHub Issues:**
#4160 (1/3) Helper methods and upgrades to other modules
#4163 (2/3) `pulse_generator` itself
#4166 (3/3) Docstring

### CODE REVIEW

**Review by lillian542 — COMMENTED** [review]
I'd like to read it in rendered form, but I'm not sure where (if anywhere) its rendered in the docs. Left a couple of suggestions/questions but will return to it tomorrow :)

File: pennylane/gradients/pulse_generator_gradient.py [hunk 1]
```diff
@@ -421,8 +420,261 @@ def processing_fn(results):
 def _pulse_generator(tape, argnum=None, shots=None, atol=1e-7):
     r"""Transform a QNode to compute the pulse generator parameter-shift gradient of pulses
     in a pulse program with respect to their inputs.
+    This method combines automatic differentiation of few-qubit operations with
+    hardware-compatible shift rules. Thus, it is limited to few-qubit pulses, but allows
+    for the evaluation of parameter-shift gradients for many-qubit pulse programs on hardware.
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:425]: The latter :) Do you have a suggestion for how to make this clearer?

File: pennylane/gradients/pulse_generator_gradient.py [hunk 2]
```diff
@@ -421,8 +420,261 @@ def processing_fn(results):
 def _pulse_generator(tape, argnum=None, shots=None, atol=1e-7):
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:420]: Oh, thanks for the pointers! I had added it to the gradients page but somehow this got lost in splitting up.
The hardware-compatible section is a very good pointer :) 

Re the first question: The last line of the file turns this function into a `gradient_transform`, inheriting the docstring. This is the reason why this private function carries the main docstring :)

File: pennylane/gradients/pulse_generator_gradient.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/gradients/pulse_generator_gradient.py:425]: I think I would flip the sentence around then and say "It allows for the evaluation of parameter-shift gradients for many-qubit pulse programs on hardware, with the limitation that the program must be composed of few-qubit pulses."

File: pennylane/gradients/pulse_generator_gradient.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/gradients/pulse_generator_gradient.py:420]: Ah, nice!

File: pennylane/gradients/pulse_generator_gradient.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:425]: Nice! :)

**Comment by dwierichs** [comment]
> I'd like to read it in rendered form, but I'm not sure where (if anywhere) its rendered in the docs. 

I think this was because I hadn't included it in the gradients section, so it was not rendered.
[Here they are now :)](https://xanaduai-pennylane--4073.com.readthedocs.build/en/4073/code/api/pennylane.gradients.pulse_generator.html) @lillian542

File: pennylane/gradients/pulse_generator_gradient.py [hunk 3]
```diff
@@ -421,8 +420,262 @@ def processing_fn(results):
...
+        before the parametrized time evolution. Here, the variable :math:`x` is just a
+        convenient way to write the modified cost function. Note that its derivative with
+        respect to :math:`x` can be computed with the standard two-term parameter-shift
+        rule for Pauli rotation gates.
+
+        *Caching*
+
+        Considering the derivation above, we notice that the same modified cost function
+        :math:`C_\ell(x)` may appear in the derivatives of distinct parameters
+        :math:`\theta_k` and :math:`\theta_m`. In order to not evaluate the same
```
Comment (frederikwilde) [inline pennylane/gradients/pulse_generator_gradient.py:670]:
```suggestion
        :math:`\theta_k` and :math:`\theta_m`, because they both enter into the same
        pulse gate. In order to not evaluate the same
```

File: pennylane/gradients/pulse_generator_gradient.py [hunk 4]
```diff
@@ -421,8 +420,262 @@ def processing_fn(results):
...
+    >>> fun(qml.execute(tapes, dev))
+    (Array(1.41897933, dtype=float64),
+     Array([0.00164914, 0.00284789], dtype=float64),
+     Array(-0.09984585, dtype=float64))
+
+    .. note::
+
+        This function requires the JAX interface and does not work with other autodiff interfaces
+        commonly encountered with PennyLane.
+        In addition, this transform is only JIT-compatible with scalar pulse parameters.
```
Comment (frederikwilde) [inline pennylane/gradients/pulse_generator_gradient.py:553]: Does scalar mean this: `params = [jnp.array(0.2), jnp.array(0.6), jnp.array(0.4)]`?
Or can it only be a single parameter?
And out of curiosity, is the caching the reason for this restriction?

**Review by frederikwilde — APPROVED** [review]
Super nice and detailed documentation! I only had some minor comments.

File: doc/introduction/interfaces.rst [hunk 5]
```diff
@@ -237,6 +237,12 @@ parameters.
 
 * ``"hadamard"``: Use hadamard tests on the generators for all compatible quantum operations arguments.
 
+* :func:`qml.gradients.stoch_pulse_grad <~.stoch_pulse_grad>`: Use a stochastic variant of the
+  parameter-shift rule for pulse programs.
+
+* :func:`qml.gradients.pulse_generator <~.pulse_generator>`: Combine automatic differentiation with
+  the parameter-shift rule for pulse programs.
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply doc/introduction/interfaces.rst:244]: I like the precision! It feels a bit technical for an intro page, what do you think? 
Maybe
```suggestion
* :func:`qml.gradients.pulse_generator <~.pulse_generator>`: Combine classical processing with the parameter-shift rule for multivariate gates to differentiate pulse programs, using effective generators.
```

File: pennylane/gradients/pulse_generator_gradient.py (hunk 4 above)
Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:553]: It's the former. I'll specify.
No, the reason for this are some assumptions at the core of the `gradients` module that surface when using JIT. Currently, operations are assumed to have parameters that all have the same shape. 
For this warning I went for the more conservative warning that all parameters need to be scalars, to be on the safe side until we update the execution/JIT interface.

Comment (dwierichs) [inline pennylane/gradients/pulse_generator_gradient.py:553]:
```suggestion
        In addition, this transform is only JIT-compatible with pulses that only have scalar parameters.
```

File: doc/introduction/interfaces.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (frederikwilde) [reply doc/introduction/interfaces.rst:244]: Even better! 💯

File: pennylane/gradients/pulse_generator_gradient.py (hunk 4 above)
Comment (frederikwilde) [reply pennylane/gradients/pulse_generator_gradient.py]: I see. JIT-ing is not always seamless.

File: pennylane/gradients/pulse_generator_gradient.py [hunk 6]
```diff
@@ -412,8 +411,263 @@ def processing_fn(results):
...
+        :math:`\tilde\omega_{k}^{(\ell)}=2i\omega_{k}^{(\ell)}`.
+        In the second to last step, we rewrote the commutator of :math:`U^\dagger BU` and
+        :math:`\frac{i}{2}P_\ell` as the derivative (at zero) of a modified cost function
+        :math:`C_\ell(x)` that executes a Pauli rotation about :math:`-i\frac{x}{2}P_\ell`
+        before the parametrized time evolution. Here, the variable :math:`x` is just a
+        convenient way to write the modified cost function. Note that its derivative with
+        respect to :math:`x` can be computed with the standard two-term parameter-shift
+        rule for Pauli rotation gates.
+
+        *Caching*
```
Comment (lillian542) [inline pennylane/gradients/pulse_generator_gradient.py:657]:
```suggestion
        **Caching**
```

File: pennylane/gradients/pulse_generator_gradient.py [hunk 7]
```diff
@@ -412,8 +411,263 @@ def processing_fn(results):
...
+
+    Executing the tapes and applying the post-processing function to the results then
+    yields the gradient:
+
+    >>> fun(qml.execute(tapes, dev))
+    (Array(1.41897933, dtype=float64),
+     Array([0.00164914, 0.00284789], dtype=float64),
+     Array(-0.09984585, dtype=float64))
+
+    .. note::
```
Comment (lillian542) [inline pennylane/gradients/pulse_generator_gradient.py:540]: I would move the note about JAX/JIT up above the example section, since I think a lot of people don't make it all the way to the end of the example, and it seems important

**Review by lillian542 — APPROVED** [review]
Looks great David, thanks for laying everything out in detail! 🎉

**Review by Qottmann — COMMENTED** [review]
These docs are amazing @dwierichs 💪 

Only one small ask to include the effective generator formula directly in the head of the doc string 🙂

File: pennylane/gradients/pulse_generator_gradient.py [hunk 8]
```diff
@@ -412,8 +411,263 @@ def processing_fn(results):
...
     in a pulse program with respect to their inputs.
+    This method combines automatic differentiation of few-qubit operations with
+    hardware-compatible shift rules.
+    It allows for the evaluation of parameter-shift gradients for many-qubit pulse programs
+    on hardware, with the limitation that the program must be composed of few-qubit pulses.
+
+    For this differentiation method, the unitary matrix corresponding to each pulse
+    is computed and differentiated with an autodiff framework. From this derivative,
+    the so-called effective generators :math:`\Omega_{k}` of the pulses
+    (one for each variational parameter) are extracted.
```
[the remark this answers is not in the document]
Comment (Qottmann) [reply pennylane/gradients/pulse_generator_gradient.py:422]: Alternatively with explicit parameters as in the theorectical background section $\frac{\partial U}{\partial \theta_k} = U(\boldsymbol{\theta}) \Omega_k$

**Review by frederikwilde — APPROVED** [review]
I think the docstring on `shots` is already very clear and unambiguous. I like the slightly more elaborate version in some of the private functions a bit more, so I'd suggest using this one for the public docstring.

**Review by Qottmann — APPROVED** [review]
great job @dwierichs 🚀 💪 👌
